### PR TITLE
CI: Add Steam build uploader

### PR DESCRIPTION
**Pull request by derrod**
### Description

Adds a GitHub actions workflow as well as necessary files for Steam build to be uploaded directly from this repo.

This is version 1 of this workflow, which does not yet account for M1/Apple Silicon builds. Once we get closer to starting the beta phase of the next release and have decided on a naming scheme for builds, whether to use universal builds for Steam, etc. the action will be updated to a "v2".

### Motivation and Context

OBS is now on Steam, we don't want to have to deal with manually uploading builds. As originally proposed in https://github.com/obsproject/rfcs/pull/30 this adds the necessary automation via GitHub actions.

Thanks to [OpenTTD](https://github.com/OpenTTD/OpenTTD) whose workflow has served an inspiration for this version.

### How Has This Been Tested?

A number of runs of various iterations on my own repo: https://github.com/derrod/obs-studio/actions/workflows/steam.yml 

A run with the version in this PR at the time of submission is here: https://github.com/derrod/obs-studio/actions/runs/2021121357

I'm sure some people have ideas to optimise it further as I'm not exactly experienced with writing github actions.

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: .github/workflows/steam.yml
Comment (PatTheMav) [inline .github/workflows/steam.yml:44]: Any reason why you don't use `apt` to install 7z? Also might be worthwhile checking for a pre-existing `7z` on the runner to skip this step.

Comment (RytoEX) [reply .github/workflows/steam.yml:44]: 7z doesn't seem to be listed in the ubuntu-20.04 image.

https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md

Comment (derrod) [reply .github/workflows/steam.yml:44]: p7zip is unmaintained and based on an ancient 7z version, so it doesn't support many newer formats like HFS+

Comment (derrod) [reply .github/workflows/steam.yml:44]: We could use this PPA instead: https://launchpad.net/~spvkgn/+archive/ubuntu/sevenzip
It has a recent enough version that would work.

**Comment by derrod** [comment]
Updated to get a more recent 7zip from a Ubuntu PPA and replaced all the ugly temporary json files with using variables that are piped into `jq`.

File: .github/workflows/steam.yml
Comment (PatTheMav) [inline .github/workflows/steam.yml:49]: I always advise to wrap variable names in curly brackets (`${EVENT}`) just as a good practise.

Not quoting the lparam is fine in bash and zsh when using `[[ ]]`. Also if you don't need substitution, I'd use single quotes here.

File: .github/workflows/steam.yml
Comment (PatTheMav) [inline .github/workflows/steam.yml:52]: Prefer `[[ ]]` on Bash, also use `==` to make it fail if used with `test` by accident.

File: .github/workflows/steam.yml
Comment (PatTheMav) [inline .github/workflows/steam.yml:88]: I'm not sure whether `-n` might suffice here - testing that only returns true if the string is set _and_ is not zero-length.

File: .github/workflows/steam.yml
Comment (PatTheMav) [inline .github/workflows/steam.yml:126]: You can `echo` the `set-output` part directly within the branch, no need to store it in an intermediary variable.

File: .github/workflows/steam.yml
Comment (PatTheMav) [inline .github/workflows/steam.yml:134]: We use `curl` above, would prefer to stick with one tool.

File: .github/workflows/steam.yml
Comment (PatTheMav) [inline .github/workflows/steam.yml:152]: Any particular reason why you use a subshell here? If you want to return to the original directory you came from, I had to kick myself to use `pushd` and `popd` for that purpose.

Otherwise I'd just use a subshell if I want to change environment variables for just a few commands.

File: .github/workflows/steam.yml
Comment (PatTheMav) [inline .github/workflows/steam.yml:172]: That is indeed the correct way to "capture" non-0 return codes and avoid a script failure. (Either the or capturing it in a variable for later handling).

**Comment by PatTheMav** [comment]
Startled the witches of nitpick forest, but looks fine overall.

Do we have to use batch scripts (i.e. prescribed by Steam)? I'd prefer Powershell, but with Windows preventing user-initiated Powershell scripts by default I wouldn't know if it works out of the box.

File: .github/workflows/steam.yml
Comment (derrod) [reply .github/workflows/steam.yml:152]: This was mostly taken from OpenTTD's steam deploy script, using the subshell with indentation just seemed like a "pretty" way of separating those things and not having to worry about where you are in the tree for subsequent commands.

**Comment by derrod** [comment]
The batch script were also simply adapted from the existing virtualcam install script. Since powershell is blocked by default that probably wouldn't work, and iirc I tried running a powershell script in an earlier iteration but that didn't work.

**Comment by PatTheMav** [comment]
> The batch script were simply adapted from the existing virtualcam install script. Since powershell is blocked by default that probably wouldn't work, and iirc I tried running a powershell script in an earlier iteration but that didn't work for some reason (maybe steam does a bit more than just telling the OS to run a file and didn't recognise `.ps1`?).

Yeah I'm worried that Windows will just block execution and batch scripts are just the way to go.

File: CI/steam/scripts/install.bat
Comment (RytoEX) [inline CI/steam/scripts/install.bat:12]: "Administrative rights are required. Please re-run this script as Administrator."

File: CI/steam/scripts/install.bat
Comment (RytoEX) [inline CI/steam/scripts/install.bat:27]: Do we want to hard-code this in this file, or use a .bat.in file and grab this GUID from CI before producing the Steam build (which is what we do for other builds, I think)?

File: CI/steam/scripts/uninstall.bat
Comment (RytoEX) [inline CI/steam/scripts/uninstall.bat:10]: "Administrative rights are required. Please re-run this script as Administrator."

File: .github/workflows/steam.yml
Comment (RytoEX) [inline .github/workflows/steam.yml:175]: Is 7zz some CLI executable?

File: .github/workflows/steam.yml
Comment (RytoEX) [inline .github/workflows/steam.yml:35]: Might as well update the first-party actions to their latest versions unless there's a reason we need the older versions.

File: .github/workflows/steam.yml
Comment (RytoEX) [inline .github/workflows/steam.yml:40]: 7zip -> 7-Zip

End with a period.

File: .github/workflows/steam.yml
Comment (RytoEX) [inline .github/workflows/steam.yml:177]: 7zip -> 7-Zip

End with a period.

**Review by RytoEX — CHANGES_REQUESTED** [review]
Did a quick review.

File: .github/workflows/steam.yml
Comment (derrod) [reply .github/workflows/steam.yml:175]: Yes, don't ask me why they went with that.

File: CI/steam/scripts/install.bat
Comment (derrod) [reply CI/steam/scripts/install.bat:27]: Could be done, though I don't see us changing it nor anybody else needing this specific script to have their own.

File: CI/steam/scripts/install.bat
Comment (derrod) [reply CI/steam/scripts/install.bat:12]: This is taken straight from https://github.com/obsproject/obs-studio/blob/master/plugins/win-dshow/virtualcam-install.bat.in#L10 so I'd suggest updating it there first.

File: CI/steam/scripts/install.bat
Comment (RytoEX) [reply CI/steam/scripts/install.bat:27]: It would just be one less place to have to consider updating it should we change it, or if we want to register the Steam build and the normal build with different GUIDs to avoid clobbering each other or having one load the other (though that may also require us to differentiate the camera names).

As mentioned, that's how we currently handle this:
* https://github.com/obsproject/obs-studio/blob/ccd1ef2bd0abb7c9d86d25871dfd8744fdd450cb/plugins/win-dshow/virtualcam-install.bat.in
* https://github.com/obsproject/obs-studio/blob/ccd1ef2bd0abb7c9d86d25871dfd8744fdd450cb/plugins/win-dshow/virtualcam-uninstall.bat.in

I'm fine with this as-is for now, but something to consider.

File: .github/workflows/steam.yml
Comment (RytoEX) [inline .github/workflows/steam.yml:19]: Do we want this to run every 24 hours or every 12 hours? The latter can be achieved with this:
```
0 */12 * * *
```

I don't have a strong opinion here.

**Review by RytoEX — APPROVED** [review]
Seems fine. I left a comment on the frequency of the cron timing. Approved otherwise.

File: .github/workflows/steam.yml
Comment (derrod) [reply .github/workflows/steam.yml:19]: I went with 24 hours so it's an actualy "night"-ly :P

**Comment by PatTheMav** [comment]
Where does Steam install the app bundle to on macOS? Did anyone test this already?

**Comment by derrod** [comment]
From memory it should be in `~/Library/Application Support/Steam/steamapps/OBS Studio`

**Comment by PatTheMav** [comment]
> From memory it should be in `~/Library/Application Support/Steam/steamapps/OBS Studio`

Ugly, but should cause no issues.

**Comment by derrod** [comment]
> > From memory it should be in `~/Library/Application Support/Steam/steamapps/OBS Studio`
> 
> Ugly, but should cause no issues.

Well it did cause the issue of browser source not loading, which was fixed in https://github.com/obsproject/obs-studio/commit/269d48f6816d77045724b050b9a6e890a301376c :P

Fun fact: Epic puts stuff into `/Users/Shared/Epic Games/<stuff>`.

**Comment by RytoEX** [comment]
@PatTheMav If this is still good to go, I'll land it.

File: .github/workflows/steam.yml
Comment (PatTheMav) [inline .github/workflows/steam.yml:52]: Use bash' `[[ ]]` here, because `==` shouldn't work with POSIX `test`.

File: .github/workflows/steam.yml
Comment (PatTheMav) [inline .github/workflows/steam.yml:59]: In bash it's advised to wrap the result of `$()` in double quotes to avoid tokenisation.

File: .github/workflows/steam.yml
Comment (PatTheMav) [inline .github/workflows/steam.yml:61]: For consistency all variable accesses should use the `${}` form.

File: .github/workflows/steam.yml
Comment (PatTheMav) [inline .github/workflows/steam.yml:72]: I'm fine with this as-is, but one should be able to do that with a single invocation of `jq`, see this example from `obs-plugintemplate`:

```
read -r deps_version deps_hash <<< \
  "$(jq -r --arg key "obs-deps" --arg target "${target}" \
    '.dependencies[$key][$target] | {version, hash} | join(" ")' \
    ${buildspec_file})"
```

You can have `jq` select the desired fields and output/join them so that `read` can extract them into separate variables.

File: .github/workflows/steam.yml
Comment (PatTheMav) [inline .github/workflows/steam.yml:103]: This is also fine as-is, I just started to prefer setting using step output variables instead of polluting the environment (especially as step outputs can also be re-used as job outputs and then also used in later jobs).

File: .github/workflows/steam.yml
Comment (PatTheMav) [inline .github/workflows/steam.yml:191]: This is just my personal choice, but when it comes to Marketplace actions that handle secret tokens, I rather specify the commit hash that I reviewed as tags can be changed after the fact and can point to unreviewed commits without anyone noticing.

File: .github/workflows/steam.yml
Comment (derrod) [reply .github/workflows/steam.yml]: I believe that was the only instance that I missed in the first round, i changed it now.

**Comment by PatTheMav** [comment]
In the best tradition of our joint indulgence of nitpicking, I added a few picked nits of my own.

File: .github/workflows/steam.yml
Comment (derrod) [reply .github/workflows/steam.yml:191]: Good point, I'll use the hashes for the steamcmd/totp action.

File: .github/workflows/steam.yml
Comment (derrod) [reply .github/workflows/steam.yml]: In this case I would disagree on the basis that the readability suffers once you get too crazy with `jq`, and given that we're talking milliseconds either way here I'm not sure it's worth changing.

File: .github/workflows/steam.yml
Comment (derrod) [reply .github/workflows/steam.yml]: I converted this to use outputs instead (although there's only one job, so reusability is kind of a moot point).

**Comment by RytoEX** [comment]
@PatTheMav Re-requesting review until we get to an approval.

**Review by PatTheMav — APPROVED** [review]

**Pull request merged**
